docs/rfc/007-integer64-type.md

```diff
@@ -0,0 +1,102 @@
+
+_The 54-bit Integer is Not Enough_
+
+Conjure currently supports three numeric types: `integer`, `double`, and `safelong`. Unfortunately for many use-cases, 2<sup>53</sup> - 1 is not large enough.
+
+These cases are relatively common and often result in one of three failure scenarios:
+1. Use safelong because it exists. We often measure nanoseconds, however a safelong can only represent three months of nanoseconds, which is long enough that it's unlikely to be caught in testing.
+1. Represent 64-bit integer values as strings in conjure. This is only as correct as the application code which sets and parses values, which is precisely the problem Conjure is built to solve. In this case the conjure definition fails to accurately describe the wire API.
+1. Use external type imports to polyfill the 64-bit integer type. This is incredibly dangerous, often resulting in code that behaves inconsistently between languages.
+
+## Proposal
+
+New conjure `integer64` type semantics matching `integer` except for the allowed values, and JSON serialized form.
+
+For the immediate future, typescript clients may represent these values using branded strings.
+Once a serialization layer is implemented for the typescript client, they may take advantage of the `integer64`
+type using the proposed [BigInt](https://github.com/tc39/proposal-bigint) type which is already supported by
+chrome, with a [polyfill](https://github.com/GoogleChromeLabs/jsbi) for other browsers.
+
+### Allowed values
+
+The `integer64` type supports signed 64-bit values ranging from -2<sup>63</sup> to 2<sup>63</sup> - 1.
+
+### JSON Format
+
+The `integer64` type is encoded to JSON as a JSON string containing the base-10 value, otherwise javascript
+clients may transparently truncate numeric values. The [BigInt](https://github.com/tc39/proposal-bigint)
+specification does not modify how json is parsed, so values must use existing JSON types.
+
+For example, take a type defined as:
+
+```yml
+types:
+  definitions:
+    objects:
+      Example:
+        fields:
+          integerField: integer
+          safelongField: safelong
+          integer64Field: integer64
+```
+
+Where each field takes the value `123`, would be serialized to json as:
+
+```json
+{
+  "integerField": 123,
+  "safelongField": 123,
+  "integer64Field": "123"
+}
+```
+
+#### The case against using JSON numeric format
+
+We use JSON numbers elsewhere in Conjure and they, by definition, better capture the integer data type. However,
+in javascript `JSON.parse` is the most common json parsing function, and it result in _incorrect data_ when
+used with numeric values beyond 52-bits. Even if we do provide a serialization layer capable of handling large
+integer values, applications are still likely to attempt usage of `JSON.parse` resulting in data loss.
+
+Secondly, many existing applications already polyfill the 64-bit integer type using string encoding with the
+following block:
+
+```yml
+types:
+  imports:
+    Long:
```

Comment:
this type definition will only work if deserializers coerce strings to longs, but I'm pretty sure that feature was disabled recently in Java, for instance.

Reply:
That is correct, the reference java implementation recently broke this, however other internal java conjure implementations still support it. My goal for rfc is to standardize behavior, and avoid maintaining multiple implementations.

My preference is to provide a supported mechanism for this feature rather than revert the coercion change.

Reply:
as Rob and I noted, it then likely needs to be handled in all languages.

I think if you weren't dealing with math, that an additional identifier type would give equal treatment in all languages and simply make more efficient use of memory in some languages than others.

Reply:
I think we are in agreement that the ecmascript bigint type is the correct binding for TS (as noted in the `Proposal` section). The full implementation requires a bit of work on our typescript generator, but I cannot invest my team in doing so until the specification change has been ratified.

Reply:
Unfortunately I don't think we could agree to this specification without the pre-work from happening in TypeScript to support the then small-ish change this RFC would make.

Reply:
To be clear: pending the typescript client serialization layer, you support this feature. Is that correct?

Reply:
Yes, if and only there is a serialization layer.